src/parse.rs

```diff
@@ -8,36 +8,65 @@ use arrayvec::ArrayVec;
 
 #[cfg(all(feature = "alloc", not(feature = "std")))]
 use crate::alloc::vec::Vec;
-use crate::error::InvalidLengthError;
+use crate::error::{InvalidLengthError, ContainsPrefixError, MissingPrefixError};
 use crate::iter::HexToBytesIter;
 
 #[rustfmt::skip]                // Keep public re-exports separate.
-pub use crate::error::{HexToBytesError, HexToArrayError};
+pub use crate::error::{FromHexError, FromNoPrefixHexError, FromPrefixedHexError, HexToArrayError, InvalidCharError};
 
 /// Trait for objects that can be deserialized from hex strings.
 pub trait FromHex: Sized {
-    /// Error type returned while parsing hex string.
-    type Error: From<HexToBytesError> + Sized + fmt::Debug + fmt::Display;
+    /// Error type returned while constructing type from byte iterator.
+    type Error: fmt::Debug + fmt::Display;
 
     /// Produces an object from a byte iterator.
     fn from_byte_iter<I>(iter: I) -> Result<Self, Self::Error>
     where
-        I: Iterator<Item = Result<u8, HexToBytesError>> + ExactSizeIterator + DoubleEndedIterator;
+        I: Iterator<Item = Result<u8, InvalidCharError>> + ExactSizeIterator + DoubleEndedIterator;
 
-    /// Produces an object from a hex string.
-    fn from_hex(s: &str) -> Result<Self, Self::Error> {
-        Self::from_byte_iter(HexToBytesIter::new(s)?)
+    /// Produces an object from a hex string that may or may not include a `0x` prefix.
+    ///
+    /// Equivalent to [`Self::from_maybe_prefixed_hex`].
+    fn from_hex(s: &str) -> Result<Self, FromHexError<Self::Error>> {
+        Self::from_maybe_prefixed_hex(s)
+    }
```

Comment:
Flagging this, this is not the functionality you talked about @Kixunil, feels to me like `from_hex` should accept 0x or not. Also, since its ambiguous I feel there is some merit to having bath short and long forms.

Reply:
The reason I wanted to go back is that most likely 99% of people who updated to our new crate got their code silently broken and they don't know about it (it wasn't in the changelog because nobody noticed). So I think we should restore it. I could also accept removing this ambiguous method entirely which forces explicit handling.

Reply:
Or maybe that isn't strong enough. Alternately we could remove `from_hex` for a couple of releases then consider re-introducing it - same aim but more aggressive.

Reply:
I just wouldn't put it back.